### ISSUE TITLE
SAVE G-CODE FILE 

### ISSUE THREAD

**Issue by arlimbad222** [gnea/grbl#993]
Hey forks 

I am making one horizontal milling & want to run only 25 line of gcode per cycle is there any way to save gcode file on board & run that file with cycle start button push and stop with stop button push. currently i am using one old laptop to send only 25 line of gcode .  and there is no editing riquired just 25 line gcode for life time .

thankyou for help.

**Comment by SimonJohnCastle** [comment]
Its a trivial exercise to store a small GCODE in an Arduino Nano and have it sent out on rest or button press.
Or get one of those cheap offline controllers and use that, if your HW supports it already.

Unfortunately you dont say what you are using.

**Comment by harryberlin** [comment]
Candle hast custom buttons for cuttom gcodes

**Comment by arlimbad222** [comment]
Hey @harryberlin 

What is candle brother I don't know about that think I only use grbl and not an coding guy basically am mechanical engineer can you please explain me breef what is candle ?

**Comment by SimonJohnCastle** [comment]
Candle is PC SW for sending GCODE to GRBL, so not what you want.

You can get a Nano clone very inexpensively or use another Uno if you have one, easy to programm, just plug in and go.
If you are new to Arduino programming, its very easy and there are many examples, all you would have to do is send the GCODE over the serial port as strings. To make things easy so you dont have to worry about the responces from GRBL you can put delays between each line to allow the machine to perform its task.

Connect GND, 5V and cross the Rx/Tx lines between the 2 Arduinos

**Comment by arlimbad222** [comment]
Hey @SimonJohnCastle 

First sorry I can't understand that things because I am mechanical guy . Suppose I have one Uno with grbl 1.1 & second Uno blank which code I have to give second uno I don't have any experience on coding if it easy can you please make one for me .

**Comment by SimonJohnCastle** [comment]
Get the Arduino IDE from [https://www.arduino.cc/en/software](https://www.arduino.cc/en/software)

The Arduino 'C' code is very simple, all you need is something like below, and upload it to your second Arduino, you can upload this and test it directly on your Uno from your PC.
If you open the Tools/Serial_Monitor you should see the output from the executing code, see image you need to select the speed 115200 which should match GRBL as they need to talk at the same bit speed (If all this is new to you)

`
void setup()
{
  Serial.begin(115200);
//  delay(10000); // Long 10 Second Delay to allow GRBL to BOOT and get ready
  Serial.println("G90");
  delay(100);
  Serial.println("G1 X10 Y10");
  delay(100);
  Serial.println("G1 X100");
  delay(1000); // Wait 1 Second for movment to complete
  Serial.println("G1 X-100");
  // ADD all your GCODE
}

void loop() { }
`

![SendGcode](https://user-images.githubusercontent.com/51445041/105397554-816ccf00-5c21-11eb-925c-5add09a202b8.jpg)

The above code will simply execute each time the Arduino resets, Press reset and see it run again in the open monitor.

**Comment by arlimbad222** [comment]
Thankyou so much @SimonJohnCastle 

Now I understand is there any way to add push button to start , hold and stop cycle for machine safety.

**Comment by SimonJohnCastle** [comment]
yes its very easy to add a button, once you see how easy it is you can learn on your own.
Look at the examples

**Comment by arlimbad222** [comment]
Thanks again @SimonJohnCastle 

I'll buy electronics tomorrow then test it if I need any help I'll drop here 🙏

**Comment by harryberlin** [comment]
check this 
https://github.com/mathertel/OneButton

**Comment by arlimbad222** [comment]
hello @harryberlin @SimonJohnCastle 

I manage  all electronic today and test the code but when i give a code to mega with grbl it gives  lots of errors (25,24,1,2,) i think i want code that waits for ok response and then give second line of gcode to arduino mega with grbl & i am not good at coding side so please give me some advice i attach my ino file please see & give me advice 

have a good day thank you all for help
[M3M5.txt](https://github.com/gnea/grbl/files/5963185/M3M5.txt)

**Comment by SimonJohnCastle** [comment]
My original idea for you, was to use a delay between each G-Code to allow the machine to move before the next command is sent.
This way you dont have to worry about the OK.
Start simple by just having one or two movements in your button press, to prove the concept. (does it work at all ?)
i.e. does Grbl get out of lock when you reset and perform the first move when you press the button.
Then add in the extra G-Code with appropriate delays
As your G-Code is a simple repeating sequence, where the movements are basically coming to a 'STOP' in any case, you can afford to delay say after the Y3, Y-25 before the X movement, so just put a long enough delay there, for each sequence of moves

**Comment by arlimbad222** [comment]
HELLO @SimonJohnCastle 

oh i just forgot to use delay between command line i will check this as soon as possible & update here .

thanks for quick responce.

**Comment by SimonJohnCastle** [comment]
you need something like an 11 second delay after the G01 Y22 F125
i.e. delay(11000);
would be interesting to know what the machine is doing ?

**Comment by arlimbad222** [comment]
hi @SimonJohnCastle 

i am using one old horizontal milling machine to make 26 tooth on cast iron ring i attach some photos .

i am using  X as a rotary axis & Y as a linear axis . not using ball screw or any cnc stuff just put 2 stepper motor and boom its working .

![IMG_20210211_173433](https://user-images.githubusercontent.com/34487708/107635128-a7214d00-6c90-11eb-998e-5e862dcdbe66.jpg)
![IMG_20210211_173423](https://user-images.githubusercontent.com/34487708/107635136-a8eb1080-6c90-11eb-9e4a-187061887d32.jpg)
![IMG_20210211_173443](https://user-images.githubusercontent.com/34487708/107635140-aab4d400-6c90-11eb-9ddc-e96c06ce6bb4.jpg)

**Comment by SimonJohnCastle** [comment]
Nice, now I understand the GCODE

**Comment by arlimbad222** [comment]
Hi @SimonJohnCastle 

Sorry for same question but I have 3 products like this & I am to lazy to find each code delay time I think there is some code that read ok message from grbl then put a second line of gcode if you have any idea how to do that please share I don't have any kind of coding experience.

Thankyou again and again .

**Comment by SimonJohnCastle** [comment]
In principle its not too difficult, assuming you dont get into error conditions and want to make it fault tolerant.
The basic idea is to read from the serial port, either into an array or String and then look for the OK.
You keep doing this until the OK arrives and then move on to sending the next GCODE,
Look up
[Serial](https://www.arduino.cc/reference/en/language/functions/communication/serial/) see readString etc
[String](https://www.arduino.cc/reference/en/language/variables/data-types/stringobject/) see lastIndexOf
I cant give you a working example of the top of my head, but its not too difficult.

Oh and put a FOR loop round your code, rather than repeating it all

The delay times are not critical, so long as they are longer than you need, and I guess 1 or 2 seconds here or there is not too impotent for you, its a matter of where you want to invest your time, learning to programm it properly may be a good investment for you

**Comment by harryberlin** [comment]
G4 should be command for waiting

**Comment by SimonJohnCastle** [comment]
This does what you need, Tested as shown, type ok and press send
Adjust it as needed for your button handling

Code attached [gcode.txt](https://github.com/gnea/grbl/files/5967353/gcode.txt)

![OK](https://user-images.githubusercontent.com/51445041/107674032-d0dd7280-6c96-11eb-999f-cd6f2440427c.jpg)

**Comment by arlimbad222** [comment]
Hi @SimonJohnCastle 

Thankyou for your efforts I will test code by tomorrow I am little busy today sorry for that I'll get back here when done something practical test .

Thankyou again and again brother .

**Comment by arlimbad222** [comment]
hi @SimonJohnCastle 

I test the code but there is little problem when i want to send some real time command like (0x91 & 0x94) by button 2 . it finish that button 1 task then button 2 is working sorry for lots of question but seriously I don't have any coding skills i just copy and pest your code  .

have a nice day my friend .
[M3M5.txt](https://github.com/gnea/grbl/files/5969890/M3M5.txt)

**Comment by SimonJohnCastle** [comment]
your string is wrong it should be "\x91" which means a single character of value hex 91.
Second I dont think GRBL responds with ok to real time commands so just use print or println like before, I dont know off my head if GRBL needs the linefeed for real time cmds, possibly not, I think it just looks in the stream for it

If your expecting to press button2 while its milling then its wont work, this is getting out of scope of what you said you initially needed, and needs more complex programming, where this is not really the place for discussion

what you have could be rearranged by the click scheduling a job in the main loop and the wait calling the tick to check for further button activity, but where will it stop

This may solve your problem but Im not familiar with the button library and have not tested it. its getting a bit reentrant 

while(Serial.available() < 2) button2.tick();

does the whole thing work correctly if you put the FOR loop in to send it 26 times ? as you removed it in your uploaded code ?

**Comment by arjunrayudu** [comment]
EXLENT SOLUTION GREAT MR. [SimonJohnCastle](https://github.com/SimonJohnCastle)
